src/main/java/frc/robot/subsystems/Launcher.java

```diff
@@ -0,0 +1,144 @@
+    lower.getPIDController().setI(Constants.LaunchConstants.launcherI);
+    lower.getPIDController().setD(Constants.LaunchConstants.launcherD);
+  }
+
+  private void updatePIDs(double p, double i, double d) {
+    upper.getPIDController().setP(p);
+    upper.getPIDController().setI(i);
+    upper.getPIDController().setD(d);
+
+    lower.getPIDController().setP(p);
+    lower.getPIDController().setI(i);
+    lower.getPIDController().setD(d);
+  }
+
+  private void updatePIDFromDashboard(String keyWord) {
+    double p = SmartDashboard.getNumber(keyWord + " P", 0);
+    double i = SmartDashboard.getNumber(keyWord + " I", 0);
+    double d = SmartDashboard.getNumber(keyWord + " D", 0);
+
+    updatePIDs(p, i, d);
+  }
+
+  public void setLaunchVelocity(double velocity) {
+    upper.getPIDController().setReference(velocity, ControlType.kSmartVelocity);
+    lower.getPIDController().setReference(velocity, ControlType.kSmartVelocity);
+  }
+
+  public Value getAnglePosition() {
+    return angleSwitcher.get();
+  }
+
+  public double getCurrentVelocity() {
+    double upperVelocity = upper.getEncoder().getVelocity();
+    double lowerVelocity = lower.getEncoder().getVelocity();
+    return (upperVelocity + lowerVelocity) / 2;
+  }
+
+  public double getVelocityDifference() {
+    double upperVelocity = upper.getEncoder().getVelocity();
+    double lowerVelocity = lower.getEncoder().getVelocity();
+    return Math.abs(upperVelocity - lowerVelocity);
+  }
+
+  public boolean isWithinVeloPercentage(double percent, double targetVelo) {
+    double currentPercent = getCurrentVelocity() / targetVelo;
+    return (1 - percent <= currentPercent && currentPercent <= 1 + percent);
+  }
+
+  public boolean differenceWithinPercentage(double percent, double targetVelo) {
+    double differencePercent = getVelocityDifference() / targetVelo;
+    return (differencePercent <= percent);
+  }
+
+  public boolean readyToLaunch(double targetVelo) {
+    boolean differenceReady =
+        differenceWithinPercentage(Constants.LaunchConstants.allowedDifferencePercent, targetVelo);
+    boolean veloReady =
+        isWithinVeloPercentage(Constants.LaunchConstants.allowedVeloPercent, targetVelo);
+    return (differenceReady && veloReady);
+  }
+
+  public void setSwitcherPosition(LaunchDistance launchDistance) {
+    switch (launchDistance) {
+      case CLOSE:
+        angleSwitcher.set(Constants.LaunchConstants.closeLaunchPosition);
+        break;
+      case FAR:
+        angleSwitcher.set(Constants.LaunchConstants.farLaunchPosition);
+        break;
+    }
+  }
+
+  public void switchAngle() {
+    boolean switcherForward = (angleSwitcher.get() == Value.kForward);
+    angleSwitcher.set(switcherForward ? Value.kReverse : Value.kForward);
+  }
+
+  public void togglePIDTuning() {
+    tuningPIDS = !tuningPIDS;
+  }
+
+  public void maintainSwitcherState() {
+    angleSwitcher.set(angleSwitcher.get());
```

Comment:
i'm pretty sure this is not right.
wont  this just set the value of the switcher to whatever it currently is?

ie , the same as doing `x = x`

Reply:
we addressed this in https://github.com/e-moser/Robot/pull/3 and i merged that.